src/track/beats.h

```diff
@@ -1,47 +1,69 @@
-#ifndef BEATS_H
-#define BEATS_H
+#pragma once
 
+#include <QByteArray>
+#include <QList>
+#include <QMutex>
 #include <QObject>
 #include <QString>
-#include <QList>
-#include <QByteArray>
-#include <QSharedPointer>
+#include <QVector>
+#include <memory>
 
-#include "util/memory.h"
+#include "proto/beats.pb.h"
+#include "track/bpm.h"
+#include "track/frame.h"
 #include "util/types.h"
 
-namespace {
-    double kMaxBpm = 500;
-}
-
 namespace mixxx {
 
 class Beats;
-typedef QSharedPointer<Beats> BeatsPointer;
+using BeatsPointer = std::shared_ptr<Beats>;
+using BeatList = QList<track::io::Beat>;
+} // namespace mixxx
 
-class BeatIterator {
-  public:
-    virtual ~BeatIterator() = default;
-    virtual bool hasNext() const = 0;
-    virtual double next() = 0;
-};
+#include "track/beatiterator.h"
+#include "track/timesignature.h"
 
-// Beats is a pure abstract base class for BPM and beat management classes. It
-// provides a specification of all methods a beat-manager class must provide, as
-// well as a capability model for representing optional features.
-class Beats : public QObject {
+class Track;
+
+namespace mixxx {
+/// Beats is a class for BPM and beat management classes.
+/// It stores beats information including beats position, down beats position,
+/// phrase beat position and changes in tempo.
+// TODO(JVC) To make it final
+// TODO(XXX): Split into 2 classes:
+// - Beats: Final, no base class, no mutex, copyable/movable
+// - BeatsObject: Derived from QObject, with mutex, not copyable/movable
+// TODO(XXX): Remove cyclic dependency on Track object, probably not necessary
+class Beats final : public QObject {
     Q_OBJECT
   public:
-    Beats() { }
+    /// Construct a BeatMap. iSampleRate may be provided if a more accurate
+    /// sample rate is known than the one associated with the Track.
+    explicit Beats(const Track* track, SINT iSampleRate = 0);
+    /// Construct a BeatMap. iSampleRate may be provided if a more accurate
+    /// sample rate is known than the one associated with the Track. If it is
+    /// zero then the track's sample rate will be used. The BeatMap will be
+    /// deserialized from the byte array.
+    Beats(const Track* track, const QByteArray& byteArray, SINT iSampleRate = 0);
+    /// Construct a BeatMap. iSampleRate may be provided if a more accurate
+    /// sample rate is known than the one associated with the Track. If it is
+    /// zero then the track's sample rate will be used. A list of beat locations
+    /// in audio frames may be provided.
+    Beats(const Track* track, const QVector<Frame>& beats, SINT iSampleRate = 0);
     ~Beats() override = default;
 
+    // TODO(JVC) Is a copy constructor needed? of we can force a move logic??
+    Beats(const Beats&);
```

Comment:
That's why I mentioned that we need to break up this class into an inner functional class that doesn't inherit from QObject and an outer QObject wrapper.

Reply:
@uklotzde How do I proceed with this?

Reply:
1. Rename `Beats` to `BeatsObject`
2. Move all required members into a plain, copyable, moveable C++ object `Beats`
  - m_subVersion
  - m_sampleRate
  - m_duration
  - m_bpm (not "m_dCachedBpm", this name is misleading)
  - m_beatList

Storing a pointer to the `Track` object should not be necessary. Instead the track should actively take control of the `BeatsObject`.

Storing the plain `Track` pointer is unsafe and may cause crashes. The only valid option would be to store a weak pointer if necessary. But try to avoid those cyclic dependencies in the first place.

Reply:
> Rename Beats to BeatsObject

Appending "Object" to a class name says nothing about what the class does. I am still in favor of "BeatVector", but I'm open to other ideas for the class name.

Reply:
I suggest the suffix "Object" to distinguish QObjects from plain C++ classes. It should only serve as a thin wrapper for signal/slot stuff. The naming of the actual class is independent of that, e.g. either `Beats` or `BeatVector`.

Reply:
We could use `BeatVector` and `BeatVectorObject` for the different abstraction levels, i.e. functional vs. infrastructure.

Reply:
I'd prefer `BeatVectorWrapper` or `BeatVectorQObject`.

Reply:
There is no need to introduce another confusing and meaningless term, then just "...QObject".

Reply:
I ran into a few problems while trying to remove the track member pointer from the Beats class.

The track pointer is currently used to dynamically find the track duration (since it can be updated later, which is a common scenario in tests)

Resolution:
1. Use TrackWeakPointer in Beats
This is semantically not wrong and could possibly work but in `Track::setBpm` a new Beats object is created and the constructor needs a shared pointer to Track itself. So weak pointer can't be used. An outer agent would need to setup the ownership between track and beats which is not doable.

2. Entirely remove track member pointer.
This would need the creation of new signals in the track class and connections would need to be established on construction to update data in Beats that can change with track.

3. Keep the current configuration
We are storing a plain pointer to track in Beats class. This means there is no ownership per se. So we don't really run into a cyclic dependency here. Track has active control of Beats and in addition to that, Beats stores a plain pointer to the track.

Reply:
I am leaning towards option 2, but I'd like to hear @uklotzde's opinion.

Reply:
Ok, I get the issue with option 1 with transforming `this` into a smart pointer. Ruled out.

Let's stick to option 3. Extending `Track` could prove difficult, because it is a very central place with lots of dependencies. Don't open the Pandora's box now ;) Later we can decide about how to move on. This does not only affect the beat grid class.

The `Track*` should only be stored in the upper level `QObject` derived class. If you need the duration or sample rate in the plain C++ class either store them as members or pass them as parameters only when needed. If they are only needed for certain operations pass them as parameters. They don't really belong to the basic concept of a *beat grid* and maintaining less state avoids potential inconsistencies.